sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala

```diff
@@ -297,6 +298,307 @@ abstract class RapidsShuffleThreadedWriterBase[K, V](
+            s"start=$start, end=$end, isLast=$doCleanUp")
+      }
+    }
+
+    // Writer task that processes partitions in order
+    val writerTask = new Runnable {
+      override def run(): Unit = {
+        var currentPartitionToWrite = 0
+        while (!processingComplete.get() || currentPartitionToWrite != maxPartitionSeen.get()) {
+          if (currentPartitionToWrite <= maxPartitionSeen.get()) {
+            // Wait for this partition's compression futures to complete
+            var containsLastForThisPartition = false
+            var futures: java.util.concurrent.CopyOnWriteArrayList[Future[(Long, Long)]] = null
+            maxPartitionSeen.synchronized {
+              futures = partitionFutures.get(currentPartitionToWrite)
+              if (currentPartitionToWrite < maxPartitionSeen.get()) {
+                containsLastForThisPartition = true
+              }
+            }
+
+            if (futures != null) {
+              // Wait for all compression tasks to complete for this partition
+              // and collect record sizes
+              import scala.collection.JavaConverters._
+              var newFutureTouched = false
+              futures.asScala.zipWithIndex.filter(pair => {
+                pair._2 >= partitionFuturesProgress.getOrDefault(currentPartitionToWrite, 0)
+              }).foreach { future =>
+                newFutureTouched = true
+                val (recordSize, compressedSize) = future._1.get()
+
+                // Write this partition
+                writePartitionBuffer(currentPartitionToWrite,
+                  partitionBytesProgress.getOrDefault(currentPartitionToWrite, 0L),
+                  partitionBytesProgress.getOrDefault(currentPartitionToWrite, 0L) +
+                    compressedSize,
+                  doCleanUp = false,
+                  partitionBuffers,
+                  mapOutputWriter)
+
+                // Update progress
+                partitionBytesProgress.put(currentPartitionToWrite,
+                  partitionBytesProgress.getOrDefault(currentPartitionToWrite, 0L) +
+                    compressedSize)
+                partitionFuturesProgress.put(currentPartitionToWrite,
+                  partitionFuturesProgress.getOrDefault(currentPartitionToWrite, 0) + 1)
+
+                // Release limiter for all records in this partition after processing is complete
+                limiter.release(recordSize)
+              }
+
+              if (containsLastForThisPartition) {
+                // just to trigger clean up
+                writePartitionBuffer(currentPartitionToWrite, 0, 0,
+                  doCleanUp = true,
+                  partitionBuffers, mapOutputWriter)
+
+                currentPartitionToWrite += 1
+              } else {
+                if (!newFutureTouched) {
+                  // Sleep briefly to avoid busy waiting
+                  Thread.sleep(1)
```

Comment:
This function is also long enough and the flow control is kind of complex. It would be good to split this up into smaller pieces so it is simple to follow.